README.md
# Arkesel SMS PHP and Laravel Package

<!-- [![Latest Version on Packagist](https://img.shields.io/packagist/v/patricpoba/arkesel-sms.svg?style=flat-square)](https://packagist.org/packages/patricpoba/arkesel-sms)
[![Build Status](https://img.shields.io/travis/patricpoba/arkesel-sms/master.svg?style=flat-square)](https://travis-ci.org/patricpoba/arkesel-sms)
[![Quality Score](https://img.shields.io/scrutinizer/g/patricpoba/arkesel-sms.svg?style=flat-square)](https://scrutinizer-ci.com/g/patricpoba/arkesel-sms)
[![Total Downloads](https://img.shields.io/packagist/dt/patricpoba/arkesel-sms.svg?style=flat-square)](https://packagist.org/packages/patricpoba/arkesel-sms) -->

This package enables sending of sms from your laravel application using https://sms.arkesel.com as a service provider.

## Requirements & Installation

This package requires at least php 7.0 or laravel 5.5. You can install the package via composer:

```bash
composer require patricpoba/arkesel-php
```

## PHP Usage

``` php
// PHP Examples

use PatricPoba\Arkesel\Sms;

$sms = new Sms('SenderId', 'smsApiKey');
  

## Basic sending(uses api_key set in .env file)
// successful response: {"code":"ok","message":"Successfully Send","balance":17706,"user":"Yaw Berko"}
// error response: {"code":"102","message":"Authentication Failed"} 
$sms->send('02XXXXXXXXX', 'Testing sms messaging');

## To use a different api key at runtime,
$sms->setApiKey('API_KEY_GOES_HERE')->send('02XXXXXXXX', 'Testing App');

## To customise sender Id (must not be more than 11 characters)
$sms->from('CompanyName')->send('02XXXXXXXX', 'Testing App');

## Sceduling (sending message at a later time)
// successful response: {"code":"109","message":"Invalid Schedule Time"} 
// successful response: {"code":"ok","message":"SMS Scheduled successfully.","balance":17705,"user":"Yaw Berko"}
$dateTime ='04-05-2020 06:19 PM'; // Must be this format - "d-m-Y h:i A" 
$sms->schedule($dateTime, '02XXXXXXXX', 'This message will be sent later')
 
## Checking Sms balance    
// successful response: {"balance":17707,"user":"Yaw Berko","country":"Ghana"}
$sms->balance();

## Check balance of a different a arkesel account account,
$sms->setApiKey('API_KEY_GOES_HERE')->balance();
```


## Laravel

If you're using laravel 5.5 and above, you can skip this step and continue at the examples.
Add the following line of code to the **providers**' array in *config/app.php file*.
```php  
PatricPoba\Arkesel\ArkeselServiceProvider::class
```

Add the facade of this package to the aliases array in the *config/app.php file*.

```php 
 'ArkeselSms' => PatricPoba\Arkesel\ArkeselSmsFacade::class
```


### Usage Examples 

``` php  
# Setting API key in .env file
Before you can start sending sms you will need to set your api key and default sender ID in your /.env file
You can find your api key here `https://sms.arkesel.com/user/sms-api/info` 
These config files can be changed  from the laravel application.  

<!-- /.env file --> 
ARKESEL_SMS_SENDER_ID=MyApp
ARKESEL_SMS_API_KEY=YourKeyGoesHere


## Sending Sms 
 
## Basic sending(uses api_key set in .env file)
 * successful response: {"code":"ok","message":"Successfully Send","balance":17706,"user":"Yaw Berko"}
 * error response: {"code":"102","message":"Authentication Failed"} 
 * */
 
ArkeselSms::send('02XXXXXXXXX', 'Testing sms messaging');

## To use a different api key at runtime,
ArkeselSms::setApiKey('API_KEY_GOES_HERE')->send('02XXXXXXXX', 'Testing App');


## To customise sender Id (must not be more than 11 characters)
ArkeselSms::from('CompanyName')->send('02XXXXXXXX', 'Testing App');


## Sceduling (sending message at a later time) 
// successful response: {"code":"109","message":"Invalid Schedule Time"} 
// successful response: {"code":"ok","message":"SMS Scheduled successfully.","balance":17705,"user":"Yaw Berko"}

$dateTime ='04-05-2020 06:19 PM'; // Must be this format - "d-m-Y h:i A" 
ArkeselSms::schedule($dateTime, '02XXXXXXXX', 'This message will be sent later')

 
## Checking Sms balance   
// successful response: {"balance":17707,"user":"Yaw Berko","country":"Ghana"}

ArkeselSms::balance();


## Check balance of a different a arkesel account account,
ArkeselSms::setApiKey('API_KEY_GOES_HERE')->balance();

```


### Security

If you discover any security related issues, please email poba.dev@outlook.com instead of using the issue tracker.


## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# Changelog

All notable changes to `arkesel-sms` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/ExampleTest.php
<?php

namespace PatricPoba\Arkesel\Tests;

use Orchestra\Testbench\TestCase;
use PatricPoba\Arkesel\ArkeselServiceProvider;

class ExampleTest extends TestCase
{

    protected function getPackageProviders($app)
    {
        return [ArkeselServiceProvider::class];
    }
    
    /** @test */
    public function true_is_true()
    {
        $this->assertTrue(true);
    }
}


tests/Unit/SmsTest.php
<?php

namespace PatricPoba\Arkesel\Tests;

use PatricPoba\Arkesel\Sms;
use Orchestra\Testbench\TestCase;

class SmsTest extends TestCase 
{
    protected $sms ;
 
	protected $testValues = [
							'senderId' => 'MyStartUp',
							'apiKey'   => 'SDERWCSXCVEARAVESARAVES'
						];
	

	function setup() : void
	{   
		$this->sms = new Sms($this->testValues['senderId'], $this->testValues['apiKey']);
    }
    

    public function test_sms_end_point_is_valid()
	{ 
		return $this->assertSame(Sms::SMS_ENDPOINT, 'https://sms.arkesel.com/sms/api');
    }
    

    public function test_new_instance_with_constructor_arguments()
	{   
		$this->assertEquals($this->testValues['senderId'], 
							$this->sms->getSenderId(), 
							'Instance var $senderId not equal to '. $this->testValues['senderId']);

		$this->assertEquals($this->testValues['apiKey'], 
							$this->sms->getApiKey(), 
							'Instance var $apiKey not equal to '. $this->testValues['apiKey']); 
	}


	public function test_contructor_arguments_can_be_overridden()
	{  
		# re-setup
		// $this->setup(); 
		$newSenderId = 'MyBusiness';		
		$this->sms->from($newSenderId); 
		$this->assertEquals($this->sms->getSenderId(), $newSenderId, 'Constructor variable $sender_id could not be overridden');

        
		$newApiKey = 'fqwerxcfqwerfxcqerf';
		$this->sms->setApiKey($newApiKey); 
		$this->assertEquals($this->sms->getApiKey(), $newApiKey, 'Constructor variable $api_key could not be overridden');
  
	}
}


tests/LaravelArkeselSmsTest.php
<?php

namespace PatricPoba\Arkesel\Tests;

use PatricPoba\Arkesel\Sms;
use Orchestra\Testbench\TestCase;
use PatricPoba\Arkesel\ArkeselServiceProvider;

class LaravelArkeselSmsTest extends TestCase
{

    protected function getPackageProviders($app)
    {
        return [ArkeselServiceProvider::class];
    }
    
    public function test_namespaces_are_valid()
	{ 
		// dd(app()->make('mNotifySms'));
		$this->assertInstanceOf(ArkeselServiceProvider::class, new ArkeselServiceProvider(app()));

		$this->assertInstanceOf(Sms::class, new Sms);

	}

	public function test_sms_facade_is_valid()
	{
		$this->assertInstanceOf(Sms::class, app()->make('arkesel-sms')); 
	}
}


src/ArkeselServiceProvider.php
<?php

namespace PatricPoba\Arkesel;

use PatricPoba\Arkesel\Sms;
use Illuminate\Support\ServiceProvider;

class ArkeselServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */ 
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('arkesel.php'),
            ], 'config'); 
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'arkesel');

        // Register the main class to use with the facade
        $this->app->singleton('arkesel-sms', function () {
            return new Sms;
        });
    }
}


src/ArkeselSmsServiceProvider.php
<?php

namespace Patricpoba\ArkeselSms;

use Illuminate\Support\ServiceProvider;

class ArkeselSmsServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'arkesel-sms');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'arkesel-sms');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('arkesel-sms.php'),
            ], 'config');

            // Publishing the views.
            /*$this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/arkesel-sms'),
            ], 'views');*/

            // Publishing assets.
            /*$this->publishes([
                __DIR__.'/../resources/assets' => public_path('vendor/arkesel-sms'),
            ], 'assets');*/

            // Publishing the translation files.
            /*$this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/arkesel-sms'),
            ], 'lang');*/

            // Registering package commands.
            // $this->commands([]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'arkesel-sms');

        // Register the main class to use with the facade
        $this->app->singleton('arkesel-sms', function () {
            return new ArkeselSms;
        });
    }
}


src/Sms.php
<?php

namespace PatricPoba\Arkesel;

class Sms
{
    /**
     * The package version. 
     * @var string
     */
    const VERSION = '1.0';

	const SMS_ENDPOINT = 'https://sms.arkesel.com/sms/api';

	protected $smsApiKey;

	public $to;

	public $message;

	public $senderId; 

	public $dateTime; 

 
    /**
	 * Although config variable $smsApiKey can be set in the env,
	 * it can also be set via the constructor to enable swapping
	 * of api keys during runtime
     *
     * @param string $senderId
     * @param string $smsApiKey
     */
	public function __construct($senderId = null, $smsApiKey = null)
	{  
        /**
         * If current app is a laravel app, fall back to system 
         * config for senderId and smsApiKey if no params are 
         * provided via constructor
         */
        if (function_exists('config')) {

            $senderId  = $senderId  ??  config('arkesel.sms.sender-id');

            $smsApiKey = $smsApiKey ??  config('arkesel.sms.api-key') ;
        }
 
        $this->from($senderId) ;

        $this->setApiKey($smsApiKey) ; 
	}
 
	/**
     * Setter for sender_id
	 * override senderId parameter
	 * @param \PatricPoba\Mnotify\Sms
	 */
	public function from($senderId)
	{
		// limit sender id to 11 characters
		$this->senderId = substr($senderId, 0, 11);

		return $this;
    }
    
    public function getSenderId()
	{
        return substr($this->senderId, 0, 11);
	}

	public function getApiKey()
	{
		return $this->smsApiKey;
	}

	public function setApiKey($smsApiKey)
	{
		// limit sender id to 11 characters
		$this->smsApiKey = $smsApiKey ;

		return $this;
	}
 
 	/**
 	 * Send Sms to Phone number
 	 * @param  int 	    $to        	recepient phone number
 	 * @param  string   $message    messge to be sent
 	 * @param  string   $key       	api key
 	 * @param  string   $dateTime 	schedule date in "d-m-Y h:i A" format
 	 * @return string   [json]
 	 */
	public function send($to = null, $message = null, $dateTime = null )
	{  
		$this->to 					= $to ?: $this->to;
		$this->message 			= $message ?: $this->message; 
		$this->dateTime 		= $dateTime ?: $this->dateTime;
 
		//prepare url
        $url = static::SMS_ENDPOINT .'?'
                    . "action=send-sms"
		            . "&api_key=". $this->getApiKey()
		            . "&to=" . $this->to 
		            . "&sms=". urlencode($this->message) 
		            . "&from=" . $this->getSenderId() ;
		            
		if( !is_null($dateTime) ) $url .= '&schedule='. urlencode($dateTime) ;
                    
        /**
         * successful response
         * {"code":"ok","message":"Successfully Send","balance":17707,"user":"Qodehub Limited"}
         * 
         * error: 
         * {"code":"102","message":"Authentication Failed"}
         */
		return $responeCode = file_get_contents($url) ; 
	}	 


    /**
     * Send sms at a later date and time.
     * @docs: inherit ->send(...) 
     * 
     * @return sample
     * {"code":"109","message":"Invalid Schedule Time"}
     * {"code":"ok","message":"SMS Scheduled successfully.","balance":17705,"user":"Qodehub Limited"}
     * @return void
     */
	public function schedule($dateTime, $to = null, $message = null )
	{
		return $this->send($to, $message, $dateTime);
	}

	/**
	 * Although config variable $smsApiKey can be set in the env,
	 * it can also be set via the constructor to enable swapping
	 * of api keys during runtime
	 * @param int $balance
	 */
	public function balance()
	{   
		$url= static::SMS_ENDPOINT .'?action=check-balance&response=json&api_key=' .$this->getApiKey(); 

		return file_get_contents($url);
	}
 

}
